src/osd/OSD.cc

```diff
@@ -9567,112 +9224,324 @@ int OSD::init_op_flags(OpRequestRef& op)
 // =============================================================
 
 #undef dout_context
-#define dout_context osd->cct
+#define dout_context cct
 #undef dout_prefix
-#define dout_prefix *_dout << "osd." << osd->whoami << " op_wq "
+#define dout_prefix *_dout << "osd." << osd->get_nodeid() << ":" << shard_id << "." << __func__ << " "
 
-void OSD::ShardedOpWQ::wake_pg_waiters(spg_t pgid)
+void OSDShard::_attach_pg(OSDShardPGSlot *slot, PG *pg)
 {
-  uint32_t shard_index = pgid.hash_to_shard(shard_list.size());
-  auto sdata = shard_list[shard_index];
-  bool queued = false;
-  unsigned pushes_to_free = 0;
+  dout(10) << pg->pg_id << " " << pg << dendl;
+  slot->pg = pg;
+  pg->osd_shard = this;
+  pg->pg_slot = slot;
+  ++osd->num_pgs;
+
+  slot->epoch = pg->get_osdmap_epoch();
+  pg_slots_by_epoch.insert(*slot);
+}
+
+void OSDShard::_detach_pg(OSDShardPGSlot *slot)
+{
+  dout(10) << slot->pg->pg_id << " " << slot->pg << dendl;
+  slot->pg->osd_shard = nullptr;
+  slot->pg->pg_slot = nullptr;
+  slot->pg = nullptr;
+  --osd->num_pgs;
```

Comment:
Short answer is because the code evolved this way.  Longer answer is that there is a lot of other state in PGSlot because the PGRef that may or may not be safe to discard at all of the call sites, and because the intention of this particular code was to make _attach_pg and _detach_pg concise and symmetric so that the invariants were clearly maintained.

Reply:
That doesn't really answer my question? Are you saying there are races on PG removal that prevent us from nuking the PGSlot right away?